### DISCUSSION TITLE
Is there something similar to "mobile: shell" but for iOS (xcrun simctl)?

### DISCUSSION THREAD

**Discussion by semeguze** [appium/appium#17675] [Q&A]
Hi! 👋 

I would like to execute some specific commands for iOS under my SauceLabs execution.

For Android I can do things like this:

```javascript
await driver.execute('mobile: shell', {command: 'settings put secure location_mode 3'});
```

And the shell command will be executed on the remote device.

I would like to do the same for this command:

```javascript
xcrun simctl location booted clear;
```

However it only works on my local machine but not with SauceLabs configuration.

I have tried using `execute` and `executeScript` commands without luck.

Any help is appreciated! 
Thanks! 🚀

**Comment by semeguze** [comment]
It looks like I already found a solution haha

By taking a look at https://github.com/appium/appium-xcuitest-driver implementation, I saw the use of this ES6/7 Node wrapper around Apple's simctl binary:

```javascript
const { Simctl } = require('node-simctl');
```

So, I was able to use the set location with the following code on my remote device:

```javascript
const simctl = new Simctl();
const allDevices = _.flatMap(_.values(await simctl.getDevices()));
const bootedDevices = allDevices.filter((device) => device.state === 'Booted');

for (const {udid} of bootedDevices) {
  simctl.udid = udid;
  log.info(`-> iOS UUID: ${udid}`);
  await simctl.setLocation(latitude, longitude);
}
```

@jlipps Would be possible to include a `mobile: command` on `appium-xcuitest-driver` to consume this `node-simctl` library similar to what https://github.com/appium/appium-uiautomator2-driver already does with `mobile: shell`?

**Reply by jlipps** [reply]
Hmm, it doesn't feel that parallel to me. `mobile: shell` lets you execute arbitrary shell scripts on an android device. I don't think we have anything similar for iOS. simctl targets a restricted set of simulator related commands, which are made accessible via regular Appium commands or mobile: commands. Can you give me an example of something you can do with simctl that you can't do with the xcuitest driver?

**Reply by semeguze** [reply]
Thanks a lot for your response!!

Well, maybe you right, they are not directly parallel because of the arbitrary nature of `adb shell` compare to the simctl subset of commands that are already covered by existing Appium commands.

However, I have only 1 scenario/example in where current Vanilla implementation was not enough to achieve my needs (or maybe I was not able to implement them in the correct way):
- I have an App which is highly dependent on GPS/Geolocation. Using the Appium's `setGeolocation` command was working when a 'realtime' update was not required, but when it was an immediate change -> validation, the following command didn't work 100% of the times:
```javascript
driver.setGeoLocation({latitude: "121.21", longitude: "11.56", altitude: "94.23"});
```
In this case, they only way I found to resolved it, was to use `forceUpdate` for Android:
```javascript
await driver.execute('mobile: execEmuConsoleCommand', {command: `geo fix ${longitude} ${latitude}`});
await driver.execute('mobile: shell', {command: `am broadcast -n ${appiumLocationReceiver} -a ${appiumLocationRetrievalAction} --ez forceUpdate true`});
```
And, xcrun simctl for iOS (I'm using simulators, so `mobile: resetLocationService` was not an option):
```javascript
const simctl = new Simctl();
const allDevices = _.flatMap(_.values(await simctl.getDevices()));
const bootedDevices = allDevices.filter((device) => device.state === 'Booted');

for (const {udid} of bootedDevices) {
  simctl.udid = udid;
  log.info(`-> iOS UUID: ${udid}`);
  await simctl.setLocation(latitude, longitude);
}
```
Only after using more "in-deep" commands I was able to achieve a faster GPS update, even if Appium offers me the `setGeolocation` command.

However, it may be something that only applied to my scenario and you may be right the fact that it is not directly equivalent with `mobile: shell`, since most of the things that `xcrun simctl` does are already covered by Appium

Thank you very much for you response! 😄

**Reply by jlipps** [reply]
hmm, if all you're trying to do is set the location, then the existing command on iOS should do that. if it doesn't, and if your use of simctl works instead, then we should just update the implementation of set location on iOS to use something that works. if the current implementation of set location on iOS doesn't work for you, mind sharing this info as a bug report?

**Reply by dayafigueroa** [reply]
> Hmm, it doesn't feel that parallel to me. `mobile: shell` lets you execute arbitrary shell scripts on an android device. I don't think we have anything similar for iOS. simctl targets a restricted set of simulator related commands, which are made accessible via regular Appium commands or mobile: commands. Can you give me an example of something you can do with simctl that you can't do with the xcuitest driver?

Deeplinking... is there any way to deeplink programmatically on ios while simulator is booted, from xcuitest?

**Reply by KazuCocoa** [reply]
Have you tried out https://appium.github.io/appium-xcuitest-driver/latest/reference/execute-methods/#mobile-deeplink ?
